[24.0] Do not copy purged outputs to object store

mvdbeek opened the pull request:
Fixes https://github.com/galaxyproject/galaxy/issues/18337. Needs a whole lot of tests still, this is just a rough pass at the most obvious cases where this might happen.

Tests to write:
- [ ] Purge output before queue (workflow and delete action ?) * outputs_to_working_directory * extra_files * dynamic outputs * extended_metadata
- [x] Purge output while running * outputs_to_working_directory * ~~extra_files~~ * dynamic outputs * extended_metadata

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek committed:
Don't copy from_work_dir outputs to purged path in shell fragment

mvdbeek committed:
Don't include purged outputs in from_work_dir outputs

mvdbeek committed:
Raise appropriate exception if require path parameter is missing

mvdbeek committed:
Don't push purged dataset contents to object store

This in particular needs a lot of new tests.
We will also need to actively purge datasets in the model store import
code, since users might have purged datasets while the job ran.
Again, more tests needed.

jmchilton commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -293,7 +293,11 @@ def __copy_if_exists_command(work_dir_output):
+    # Users can purge outputs before the job completes,
+    # in that case we don't want to copy the output to a purged path.
+    # Static, non work_dir_output files are handled in job_finish code.
+    return f'\nif [ -f "{source_file}" -a -f "{destination}" ] ; then cp "{source_file}" "{destination}" ; fi'
```
This seems to be a wild check 😭. I hate it but I don't know the alternative. I know people love the disk object store and jobs writing directly to the object store 🤔.

mvdbeek replied:
It's a bit of an "optimization" .. I'm happy to restore the original version, the important thing is that we don't store data in purged datasets.

mvdbeek committed:
Add tool that produces all types of outputs

mvdbeek committed:
Don't fail _finish_or_resubmit if tool_stdout / tool_stderr not written

These are only written if the command actually ran, so we'd fail here
for instance if the outputs are deleted before the job had a chance to
run.

mvdbeek committed:
Fix up test cases

Add a sleep so we can delay output purging until command line is
templated. If we don't sleep we generate a command line where the
output path is just `''`. That needs another fix!

mvdbeek committed:
Fix handling of collecting discovered but purged outputs

mvdbeek committed:
Elevate external metadata setting to warning

This is a tricky one. There could be systematic issues here that an
admin would want to fix, but it could also just be the case that a user
forced a wrong datatype. Ideally we'd probably tag this as job execution
issue ? It's not very different from a job error IMO.

mvdbeek committed:
Fix metadata setting on purged outputs

mvdbeek committed:
Extend test to make sure disk path remains purged

mvdbeek committed:
Set file_size and total_size to 0 for discovered but purged datasets

mvdbeek committed:
Purge dataset if finishing job for purged output

mvdbeek committed:
Do not overwrite purged or deleted states when importing datasets from job

mvdbeek committed:
Delete purged outputs from object store when importing dataset for purged output

mvdbeek committed:
Fix up tests
